nidx/tests/test_searcher_cluster.rs

```diff
@@ -0,0 +1,245 @@
+                    "english".to_string(),
+                    VectorIndexConfig {
+                        vector_dimension: Some(3),
+                        ..Default::default()
+                    },
+                )]),
+                ..Default::default()
+            }))
+            .await?;
+        shards.push(uuid::Uuid::parse_str(&response.into_inner().id).unwrap());
+    }
+
+    // Create a cluster with 3 nodes
+    let nodes = Arc::new(Mutex::new(Vec::new()));
+    let mut searchers = Vec::new();
+    for _ in 0..3 {
+        let searcher_server = GrpcServer::new("localhost:0").await?;
+        let searcher_port = searcher_server.port()?;
+        nodes.lock().unwrap().push(format!("localhost:{searcher_port}"));
+        let list_nodes = ManualListNodes(nodes.clone(), format!("localhost:{searcher_port}"));
+        let work_dir = tempdir()?;
+        let searcher = SyncedSearcher::new(fixture.settings.metadata.clone(), work_dir.path());
+        let arc_list_nodes = Arc::new(list_nodes);
+        let searcher_api = SearchServer::new(searcher.index_cache(), ShardSelector::new(arc_list_nodes.clone(), 1));
+        searchers.push(format!("localhost:{searcher_port}"));
+        let settings_copy = fixture.settings.clone();
+        let shutdown = CancellationToken::new();
+        tokio::task::spawn(searcher_server.serve(searcher_api.into_service(), shutdown.clone()));
+        tokio::task::spawn(async move {
+            searcher
+                .run(
+                    settings_copy.storage.as_ref().unwrap().object_store.clone(),
+                    settings_copy.searcher.clone().unwrap_or_default(),
+                    shutdown.clone(),
+                    ShardSelector::new(arc_list_nodes, 1),
+                    None,
+                )
+                .await
+        });
+    }
+    tokio::time::sleep(Duration::from_secs(1)).await;
+
+    // All shards are accessible from all nodes
+    for shard in &shards {
+        for searcher in &searchers {
+            NidxSearcherClient::connect(format!("http://{searcher}"))
+                .await?
+                .search(Request::new(SearchRequest {
+                    shard: shard.to_string(),
+                    ..Default::default()
+                }))
+                .await?;
+        }
+    }
+
+    // Remove one node from the cluster and wait for sync, searches should still work
+    nodes.lock().unwrap().remove(2);
+    tokio::time::sleep(Duration::from_secs_f32(1.2)).await;
+    for shard in &shards {
+        for searcher in &searchers[0..2] {
+            NidxSearcherClient::connect(format!("http://{searcher}"))
+                .await?
+                .search(Request::new(SearchRequest {
+                    shard: shard.to_string(),
+                    ..Default::default()
+                }))
+                .await?;
+        }
+    }
+
+    Ok(())
+}
+
+#[sqlx::test]
+async fn test_search_cluster_shard_distribution(pool: PgPool) -> anyhow::Result<()> {
```

Comment:
Well, it's very unlikely that this happens since shard distribution is random. But maybe a quicker solution is to add an assert in the first pass to make sure that some shards were in the node 2? I can also assert that the rest of the shards were not moved, only the ones in shard 1 were :thinking:

Reply:
I added a bunch more asserts to test that shards are moving around as intended.